src/sig_stfl/sig_stfl.c

```diff
@@ -0,0 +1,1435 @@
+#ifdef OQS_ENABLE_SIG_STFL_lms_sha256_h10_w8_h5_w8
+		return OQS_SECRET_KEY_LMS_SHA256_H10_W8_H5_W8_new();
+#else
+		return NULL;
+#endif
+	} else if (0 == strcasecmp(method_name, OQS_SIG_STFL_alg_lms_sha256_h10_w2_h10_w2)) {
+#ifdef OQS_ENABLE_SIG_STFL_lms_sha256_h10_w2_h10_w2
+		return OQS_SECRET_KEY_LMS_SHA256_H10_W2_H10_W2_new();
+#else
+		return NULL;
+#endif
+	} else if (0 == strcasecmp(method_name, OQS_SIG_STFL_alg_lms_sha256_h10_w4_h10_w4)) {
+#ifdef OQS_ENABLE_SIG_STFL_lms_sha256_h10_w4_h10_w4
+		return OQS_SECRET_KEY_LMS_SHA256_H10_W4_H10_W4_new();
+#else
+		return NULL;
+#endif
+	} else if (0 == strcasecmp(method_name, OQS_SIG_STFL_alg_lms_sha256_h10_w4_h5_w8)) {
+#ifdef OQS_ENABLE_SIG_STFL_lms_sha256_h10_w4_h5_w8
+		return OQS_SECRET_KEY_LMS_SHA256_H10_W4_H5_W8_new();
+#else
+		return NULL;
+#endif
+	} else if (0 == strcasecmp(method_name, OQS_SIG_STFL_alg_lms_sha256_h10_w8_h10_w8)) {
+#ifdef OQS_ENABLE_SIG_STFL_lms_sha256_h10_w8_h10_w8
+		return OQS_SECRET_KEY_LMS_SHA256_H10_W8_H10_W8_new();
+#else
+		return NULL;
+#endif
+	} else if (0 == strcasecmp(method_name, OQS_SIG_STFL_alg_lms_sha256_h15_w8_h5_w8)) {
+#ifdef OQS_ENABLE_SIG_STFL_lms_sha256_h15_w8_h5_w8
+		return OQS_SECRET_KEY_LMS_SHA256_H15_W8_H5_W8_new();
+#else
+		return NULL;
+#endif
+	} else if (0 == strcasecmp(method_name, OQS_SIG_STFL_alg_lms_sha256_h15_w8_h10_w8)) {
+#ifdef OQS_ENABLE_SIG_STFL_lms_sha256_h10_w8_h10_w8
+		return OQS_SECRET_KEY_LMS_SHA256_H15_W8_H10_W8_new();
+#else
+		return NULL;
+#endif
+	} else if (0 == strcasecmp(method_name, OQS_SIG_STFL_alg_lms_sha256_h15_w8_h15_w8)) {
+#ifdef OQS_ENABLE_SIG_STFL_lms_sha256_h15_w8_h15_w8
+		return OQS_SECRET_KEY_LMS_SHA256_H15_W8_H15_W8_new();
+#else
+		return NULL;
+#endif
+	} else if (0 == strcasecmp(method_name, OQS_SIG_STFL_alg_lms_sha256_h20_w8_h5_w8)) {
+#ifdef OQS_ENABLE_SIG_STFL_lms_sha256_h20_w8_h5_w8
+		return OQS_SECRET_KEY_LMS_SHA256_H20_W8_H5_W8_new();
+#else
+		return NULL;
+#endif
+	} else if (0 == strcasecmp(method_name, OQS_SIG_STFL_alg_lms_sha256_h20_w8_h10_w8)) {
+#ifdef OQS_ENABLE_SIG_STFL_lms_sha256_h20_w8_h10_w8
+		return OQS_SECRET_KEY_LMS_SHA256_H20_W8_H10_W8_new();
+#else
+		return NULL;
+#endif
+	} else if (0 == strcasecmp(method_name, OQS_SIG_STFL_alg_lms_sha256_h20_w8_h15_w8)) {
+#ifdef OQS_ENABLE_SIG_STFL_lms_sha256_h20_w8_h15_w8
+		return OQS_SECRET_KEY_LMS_SHA256_H20_W8_H15_W8_new();
+#else
+		return NULL;
+#endif
+	} else if (0 == strcasecmp(method_name, OQS_SIG_STFL_alg_lms_sha256_h20_w8_h20_w8)) {
+#ifdef OQS_ENABLE_SIG_STFL_lms_sha256_h20_w8_h20_w8
+		return OQS_SECRET_KEY_LMS_SHA256_H20_W8_H20_W8_new();
+#else
+		return NULL;
+#endif
+	}
+//#endif //OQS_ENABLE_SIG_STFL_LMS
+	else {
+		return NULL;
+	}
+}
+
+OQS_API void OQS_SIG_STFL_SECRET_KEY_free(OQS_SIG_STFL_SECRET_KEY *sk) {
+	if (sk == NULL || sk->free_key == NULL) {
+		return;
+	}
+
+	/* Call object specific free */
+	sk->free_key(sk);
+
+	/* Free sk object */
+	OQS_MEM_secure_free(sk, sizeof(sk));
+	sk = NULL;
+}
```

Comment:
It's my bad. I said "freed" but in actuality it had an issue with zeroing things out. :)

Reply:
can you suggest new changes? I'm confused what need to be done here.

Reply:
So currently if I want to deserialize a secret key I need to give it an `OQS_SIG_STFL *` argument. If I want that struct to live for awhile, I would probably do something like `OQS_SIG_STFL *sig = malloc(sizeof(*sig));`. However, let's say that `deserialize` fails for some reason. Intuitively I could just call `free(sig)`, however I feel like it should also be valid to call `OQS_SIG_STFL_free(sig)`. With how it's implemented currently, it would not free the struct unless `free_key` was not null (or in other words deserialize succeeded).

My suggestions above ensures that the `OQS_SIG_STFL` being pointed to is always freed, even if it's not initialized. By changing the argument from a single pointer to a double pointer, this allows us to consume the `OQS_SIG_STFL *` and once `OQS_SIG_STFL_free` returns, ensures that it points to `NULL` instead of its old address.

Line 1351 in the existing implementation is essentially a no-op. Here's an example!
![Screenshot 2024-04-16 at 8 38 21 PM](https://github.com/open-quantum-safe/liboqs/assets/6541356/6ec30076-e506-4bc4-8f0f-49e9ecd2d564)

Reply:
To address the problem, "what if `free_key` is not assigned or NULL?"

It's assigned at the struct creation. 

In `sig_stfl_xmss_secret_key_functions.c`: 

```c
// Set Secret Key free function
sk->free_key = OQS_SECRET_KEY_XMSS_free;
```

It's assigned a specific function to clean up XMSS key. 
Similarly, at `sig_stfl_lms.c`: 

```c
sk->free_key = OQS_SECRET_KEY_LMS_free;
```
To conclude, `free_key` is assigned at the creation of struct `OQS_SIG_STFL_SECRET_KEY`. 
So when `OQS_SIG_STFL_free(x)` is called, it cleans up all the underlying data in the struct and the struct itself. 
The `free()` method wouldn't work, simply this struct is more complicated than a simple `free()`. 

Do you agree? We can talk about the use of double pointer later.  


```
If I want that struct to live for awhile, I would probably do something like OQS_SIG_STFL *sig = malloc(sizeof(*sig));. However, let's say that deserialize fails for some reason. Intuitively I could just call free(sig), however I feel like it should also be valid to call OQS_SIG_STFL_free(sig). 
```

How about you create a new SIG_STFL object by `OQS_SIG_STFL_SECRET_KEY_new` instead of plain `malloc(sizeof(x))`?

Reply:
Yep that resolves my first comment :-) (it also fixed a bug I was having so thanks!).

However we should either remove the sk = NULL line as it is a redundant operation and it could leave people skimming the code to believe that the pointer is indeed being set to NULL when it isn't. We should probably double check this isn't an issue in other frees as well.

Reply:
When the key is proper cleaned up by `OQS_SIG_STFL_free(sig)`, the function should also nullify the `sig` pointer. 
I prefer this convention, whenever I have struct to free, my convention is always `free_struct(s)` and `s` become `NULL` after `free_struct()` executed. 

I against removing `sk=NULL` in `OQS_SIG_STFL_SECRET_KEY_free(sk)`, for the same reason above. I can be wrong here. What is the ideal approach, if one had freed the secret key, and it's not assigned to NULL?

```c
OQS_SIG_STFL_SECRET_KEY_free(sk); // sk is not assigned to NULL

/* 
This check will pass:
if (sk == NULL || sk->free_key == NULL) {
		return;
	}
*/

OQS_SIG_STFL_SECRET_KEY_free((sk);  // Double free at OQS_MEM_secure_free(sk, sizeof(sk));

```

Reply:
If you look at the screenshot above you can see that a function that takes a single pointer and sets that pointer to NULL doesn't do anything. When the function returns to the parent scope the pointer that was passed in will still point to the same thing it did before the function call. The double pointer is how you can get the pointer to point to NULL once it returns to the parent scope. :-)

So if you want the behaviour you described above it needs to be a double pointer.

Reply:
My vote would go to keeping the single pointer and removing the `sk = NULL` assignment in the function (as it's indeed basically a no-op). In my opinion, it's up to the programmer using the API to `NULL`-set pointers after calling any sort of "free" function, just as with the standard C `free` function, and a single-pointer function signature is is consistent with `OQS_SIG_free` (and other similar functions in liboqs).

If you're both OK with this, @Martyrshot and @cothan, I can go ahead and add a commit to this branch removing the `sk = NULL` assignment and fixing the size passed to `OQS_MEM_secure_free` in the same function. No sense in making a PR for such a small change.

Reply:
Sounds good to me. :)

I haven't had a chance to check other places for the size issue, so sorry to ask for a bit more work, but would you mind doing a quick sanity check on other places `OQS_MEM_secure free` is called in the `sig_stfl` codebase? If not, I will try to look later tonight

Reply:
> I haven't had a chance to check other places for the size issue, so sorry to ask for a bit more work, but would you mind doing a quick sanity check on other places `OQS_MEM_secure free` is called in the `sig_stfl` codebase? If not, I will try to look later tonight

I did a `grep` on `OQS_MEM_secure_free` when you first pointed out the issue, and I believe this is the only place where the issue occurs :)